internal/http/websocket/v2/websocket.go

```diff
@@ -0,0 +1,255 @@
+	"time"
+
+	"github.com/gobwas/ws"
+	"github.com/gobwas/ws/wsutil"
+)
+
+const (
+	// Time allowed to write a message to the peer.
+	writeWait = 10 * time.Second
+	// Time allowed to read the next pong message from the peer.
+	pongWait   = 60 * time.Second
+	pingPeriod = (pongWait * 9) / 10
+)
+
+func read(conn *connHander, cli *Client) {
+	defer func() {
+		cli.d <- conn
+		conn.rwc.Close()
+	}()
+
+	conn.setReadDeadLine(pongWait)
+
+	log.Printf("read: %v\n", conn)
+
+	for {
+		msg, err := conn.read()
+		if err != nil {
+			// handle error
+			log.Printf("read err: %v\n", err)
+			break
+		}
+
+		log.Printf("read msg: %v\n", msg)
+
+		cli.bc <- msg
+	}
+}
+
+func write(conn *connHander) {
+	ticker := time.NewTicker(pingPeriod)
+	defer func() {
+		ticker.Stop()
+		conn.rwc.Close()
+	}()
+
+	log.Printf("write: %v\n", conn)
+
+	for {
+		select {
+		case msg, ok := <-conn.send:
+			conn.setWriteDeadLine(writeWait)
+			if !ok {
+				conn.write(&wsutil.Message{OpCode: ws.OpClose, Payload: []byte{}})
+				return
+			}
+
+			if err := conn.write(msg); err != nil {
+				conn.logf("msg err: %v\n", err)
+				return
+			}
+		case <-ticker.C:
+			conn.setWriteDeadLine(writeWait)
+			if err := conn.write(&wsutil.Message{OpCode: ws.OpPing, Payload: nil}); err != nil {
+				conn.logf("ticker err: %v\n", err)
+				return
+			}
+		}
+	}
+}
+
+type Client struct {
+	r, d chan *connHander
+	bc   chan *wsutil.Message
+	cs   map[*connHander]bool
+	u    *ws.HTTPUpgrader
+
+	// Capacity of the send channel.
+	// If capacity is 0, the send channel is unbuffered.
+	Capacity uint
+}
+
+// Get the number of connections
+func (cli *Client) Len() int {
+	// NOTE a mutex may or may not be required
+	// cli.lock.Lock()
+	// defer cli.lock.Unlock()
+	return len(cli.cs)
+}
+
+// TODO -- should be able to close all connections via their own channels
+func (cli *Client) Close() error {
+	defer func() {
+		// close channels
+		close(cli.r)
+		close(cli.d)
+		close(cli.bc)
+	}()
+
+	cli.bc <- &wsutil.Message{OpCode: ws.OpClose, Payload: []byte{}} // broadcast close
+	return nil
+}
+
+func NewClient(cap uint) *Client {
+	cli := &Client{
+		r:  make(chan *connHander),
+		d:  make(chan *connHander),
+		bc: make(chan *wsutil.Message),
+		cs: make(map[*connHander]bool),
+		u:  &ws.HTTPUpgrader{
+			// TODO may be fields here that worth setting
+		},
+		Capacity: cap,
+	}
+
+	go cli.listen()
+	return cli
+}
+
+func (cli *Client) listen() {
+	for {
+		select {
+		case conn := <-cli.r:
+			cli.cs[conn] = true
+		case conn := <-cli.d:
+			delete(cli.cs, conn)
+			close(conn.send)
+		case msg := <-cli.bc:
+			for conn := range cli.cs {
+				select {
+				case conn.send <- msg:
+				default:
+					close(conn.send)
+					delete(cli.cs, conn)
+				}
+			}
+		}
+	}
+}
+
+func (cli *Client) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// TODO check capacity
+	if cli.Capacity > 0 && len(cli.cs) >= int(cli.Capacity) {
+		http.Error(w, "too many connections", http.StatusTooManyRequests)
```

Comment:
I'm not sure `StatusTooManyRequests` is the proper response here. That [code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429) is usually used in response to a client spamming a service with too many requests in a given time period. It's not really the client's fault if the room is already full, they just happened to hit it at the wrong time. I'm not sure what the best alternative code is, but maybe [503 - Service Unavailable](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/503) is better? 
```suggestion
		http.Error(w, "too many connections", http.StatusTooManyRequests)
```

Reply:
ah ok I can change to a 503, was very unsure about codes for that.